### PR TITLE
Fix tokens comparison in Payment for temREDUNDANT error

### PR DESCRIPTION
**Pull request by gregtatcam**
## High Level Overview of Change

Fix semantics of tokens comparison in the payment transactor when checking for temREDUNDANT.
This check prevents an account sending tokens to self.
Current code checks for equal tokens (currency or MPTID) AND the issuers. It should only check
if currency or MPTID are equal.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### CODE REVIEW

**Comment by sgramkumar** [comment]
This patch fixes the test failure 👍

File: src/test/app/MPToken_test.cpp [hunk 1]
```diff
@@ -1938,6 +1938,124 @@ class MPToken_test : public beast::unit_test::suite
...
+            BEAST_EXPECT((mptAmt1 -= mptAmt2) == mptAmt1);
+            BEAST_EXPECT(mptAmt1 == mptAmt2);
+            BEAST_EXPECT(mptAmt1 == 100);
+            BEAST_EXPECT(MPTAmount::minPositiveAmount() == MPTAmount{1});
+        }
+
+        {
+            testcase("Test MPTIssue from/to Json");
+            MPTIssue const issue1{asset1.get<MPTIssue>()};
+            Json::Value const jv = to_json(issue1);
```
Comment (kennyzlei) [inline src/test/app/MPToken_test.cpp:2046]: nit: can we explicitly check what the contents of `jv` are here?  Otherwise we just have confidence that `to_json` and `mptIssueFromJson` are inverse functions without knowing if the json `jv` was properly created

File: src/test/app/MPToken_test.cpp [hunk 2]
```diff
@@ -1938,6 +1938,124 @@ class MPToken_test : public beast::unit_test::suite
...
+            MPTIssue const issue2 = mptIssueFromJson(jv);
+            BEAST_EXPECT(issue1 == issue2);
+        }
+
+        {
+            testcase("Test Asset from/to Json");
+            Json::Value jv;
```
Comment (kennyzlei) [inline src/test/app/MPToken_test.cpp:2053]: nit: I think it would be more thorough to test more than just an empty json value here

**Review by kennyzlei — APPROVED** [review]
Unit tests lgtm

File: include/xrpl/protocol/Asset.h [hunk 3]
```diff
@@ -97,6 +97,13 @@ class Asset
 
     friend constexpr bool
     operator==(Currency const& lhs, Asset const& rhs);
+
+    /** Return true if both asset's issue (Issue or MPTIssue)
+     * are the same and hold the same token (currency or MPTID).
+     * Otherwise return false.
```
Comment (thejohnfreeman) [inline include/xrpl/protocol/Asset.h:103]: I'm suggesting a change to the whole string, but wanted to note the original string was one of the rare occasions you'd need the plural possessive "assets'" (apostrophe _after_ the s). You'd need to pluralize "issues" too.

```suggestion
    /** Return true if both assets refer to the same currency (regardless of issuer) or MPT issuance.
     * Otherwise return false.
```

File: src/test/app/MPToken_test.cpp [hunk 4]
```diff
@@ -1938,6 +1938,126 @@ class MPToken_test : public beast::unit_test::suite
...
+            STAmount mptOverflow{asset2, 3037000500ull};
+            try
+            {
+                res = multiply(mptOverflow, mptOverflow, asset3);
+                fail("should throw runtime exception 1");
+            }
+            catch (std::runtime_error const&)
+            {
+                pass();
+            }
```
Comment (ximinez) [inline src/test/app/MPToken_test.cpp:2017]: Capture the error here and check that the `what()` is what you're expecting.
```suggestion
            catch (std::runtime_error const& e)
            {
                BEAST_EXPECTS(e.what() == "MPT value overflow", e.what());
            }
```

File: src/test/app/MPToken_test.cpp [hunk 5]
```diff
@@ -1938,6 +1938,126 @@ class MPToken_test : public beast::unit_test::suite
...
+            try
+            {
+                res = multiply(
+                    STAmount{asset1, (2ull << 32) + 2}, mptOverflow, asset3);
+                fail("should throw runtime exception 2");
+            }
+            catch (std::runtime_error const&)
+            {
+                pass();
+            }
```
Comment (ximinez) [inline src/test/app/MPToken_test.cpp:2029]: Same here.
```suggestion
            catch (std::runtime_error const& e)
            {
                BEAST_EXPECTS(e.what() == "MPT value overflow", e.what());
            }
```

File: src/test/app/MPToken_test.cpp [hunk 6]
```diff
@@ -1938,6 +1938,126 @@ class MPToken_test : public beast::unit_test::suite
...
+        }
+
+        {
+            testcase("Test MPTAmount arithmetics");
+            MPTAmount mptAmt1{100};
+            MPTAmount const mptAmt2{100};
+            BEAST_EXPECT((mptAmt1 += mptAmt2) == MPTAmount{200});
+            BEAST_EXPECT((mptAmt1 -= mptAmt2) == mptAmt1);
```
Comment (ximinez) [inline src/test/app/MPToken_test.cpp:2037]: If the `+=` and `-=` operators are badly messed up, and return the summed value without actually modifying the amount, these tests will still pass. Even though that's far fetched, it doesn't hurt to check the value result explicitly.
```suggestion
            BEAST_EXPECT((mptAmt1 += mptAmt2) == MPTAmount{200});
            BEAST_EXPECT(mptAmt1 == 200);
            BEAST_EXPECT((mptAmt1 -= mptAmt2) == mptAmt1);
```

File: src/test/app/MPToken_test.cpp [hunk 7]
```diff
@@ -1938,6 +1938,126 @@ class MPToken_test : public beast::unit_test::suite
...
+                jv[jss::mpt_issuance_id] == to_string(asset1.get<MPTIssue>()));
+            BEAST_EXPECT(issue1 == mptIssueFromJson(jv));
+        }
+
+        {
+            testcase("Test Asset from/to Json");
+            Json::Value const jv = to_json(asset1);
+            BEAST_EXPECT(
+                jv[jss::mpt_issuance_id] == to_string(asset1.get<MPTIssue>()));
+            BEAST_EXPECT(asset1 == assetFromJson(jv));
```
Comment (ximinez) [inline src/test/app/MPToken_test.cpp:2057]: Could we also explicitly check the contents of `jv` here?

**Review by ximinez — CHANGES_REQUESTED** [review]

**Comment by ximinez** [comment]
Also, I notice the linux CI builds are failing.

File: include/xrpl/protocol/Asset.h [hunk 8]
```diff
@@ -97,6 +97,12 @@ class Asset
 
     friend constexpr bool
     operator==(Currency const& lhs, Asset const& rhs);
+
+    /** Return true if both assets refer to the same currency (regardless of
+     * issuer) or MPT issuance. Otherwise return false.
+     */
+    friend constexpr bool
+    equalTokens(Asset const& lhs, Asset const& rhs);
```
[the remark this answers is not in the document]
Comment (gregtatcam) [reply include/xrpl/protocol/Asset.h:105]: Well, MPT is not a currency. I thought that `token` better reflects both Currency and MPT.

[the remark this answers is not in the document]
Comment (ximinez) [reply include/xrpl/protocol/Asset.h:105]: Fair enough.

File: src/test/app/MPToken_test.cpp (hunk 7 above)
Comment (gregtatcam) [reply src/test/app/MPToken_test.cpp:2057]: I'm not following. We checked above for `jv[jss::mpt_issuance_id]`.

**Comment by gregtatcam** [comment]
> Also, I notice the linux CI builds are failing.

Thanks, I'm investigating.

**Review by ximinez — APPROVED** [review]

**Review by thejohnfreeman — APPROVED** [review]
